src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,351 @@
+
+  coveragerc_digest = await Get[Digest](InputFilesContent, get_coveragerc_input(DEFAULT_COVERAGE_CONFIG))
+
+  coverage_directory_digests = await MultiGet(
+    Get[Digest](
+      DirectoryWithPrefixToAdd(
+        directory_digest=result.test_result._python_sqlite_coverage_file,
+        prefix=result.address.to_address().path_safe_spec,
+      )
+    )
+    for result in test_results
+    if result.test_result is not None
+    and result.test_result._python_sqlite_coverage_file is not None
+  )
+  sources_digests = [
+    hydrated_target.adaptor.sources.snapshot.directory_digest
+    for hydrated_target in transitive_targets.closure
+    if hasattr(hydrated_target.adaptor, 'sources')
+  ]
+
+  merged_input_files: Digest = await Get(
+    Digest,
+    DirectoriesToMerge(directories=(
+      *coverage_directory_digests,
+      *sources_digests,
+      coveragerc_digest,
+      coverage_setup.requirements_pex.directory_digest,
+    )),
+  )
+
+  prefixes = [f'{result.address.to_address().path_safe_spec}/.coverage' for result in test_results]
+  coverage_args = ['combine', *prefixes]
+  request = coverage_setup.requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path=f'./{coverage_setup.requirements_pex.output_filename}',
+    pex_args=coverage_args,
+    input_files=merged_input_files,
+    output_files=('.coverage',),
+    description=f'Merge coverage reports.',
+  )
+
+  result = await Get[ExecuteProcessResult](
+    ExecuteProcessRequest,
+    request
+  )
+  return MergedCoverageData(coverage_data=result.output_directory_digest)
+
+
+def get_file_names(all_target_adaptors):
+  # TODO(#4535): This functionality should be provided.
+  def iter_files():
+    for adaptor in all_target_adaptors:
+      if hasattr(adaptor, 'sources'):
+        for file in adaptor.sources.snapshot.files:
+          if file.endswith('.py'):
+            yield file
+
+  return list(iter_files())
+
+
+@dataclass(frozen=True)
+class PytestCoverageReport():
+  report_directory_digest: Digest
+  directory_to_materialize_to: PurePath
+
+
+@rule(name="Generate coverage report")
+async def generate_coverage_report(
+  transitive_targets: TransitiveHydratedTargets,
+  python_setup: PythonSetup,
+  coverage_setup: CoverageSetup,
+  merged_coverage_data: MergedCoverageData,
+  coverage_toolbase: PytestCoverage,
+  source_root_config: SourceRootConfig,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+) -> PytestCoverageReport:
+  """Takes all python test results and generates a single coverage report."""
+  requirements_pex = coverage_setup.requirements_pex
+  python_targets = [
+    target for target in transitive_targets.closure
+    if target.adaptor.type_alias == 'python_library' or target.adaptor.type_alias == 'python_tests'
+  ]
+
+  source_roots = source_root_config.get_source_roots()
+  python_files = frozenset(itertools.chain.from_iterable(
```

Comment:
Does `frozenset` provide consistent ordering guarantees? If not, we should use an `OrderedSet` or similar because this is used to generate a file which is an input to a process execution, and we want consistent cache keys

Reply:
frozenset does not provide ordering guarantees, nice catch! This value though is used only in a function call (not a Get) and it is sorted before we write it into the coverage config (which is cached as part of the coveragerc_digest.) See line 108 for where it's used.